shell/src/context_listener.rs

```diff
@@ -118,35 +126,77 @@ fn listen_protocol_events(
+                        }
+
+                        context_storage.put(&block_hash.clone(), &ContextRecordValue::new(msg))?;
+                    }
+                    ContextAction::Delete { block_hash: Some(block_hash), key, .. }
+                    | ContextAction::RemoveRecord { block_hash: Some(block_hash), key, .. } => {
+                        let state = get_default(&mut blocks, block_hash.clone());
+                        state.insert(key.join("/"), Bucket::Deleted);
+
+                        context_storage.put(&block_hash.clone(), &ContextRecordValue::new(msg))?;
+                    }
+                    ContextAction::Commit { new_context_hash, block_hash: Some(block_hash), .. } => {
+                        if let Some(block) = blocks.get(block_hash) {
+                            state.merge(block);
+                        } else {
+                            warn!(log, "Trying to commit non-existent block"; "block" => format!("{:?}", block_hash));
+                        }
+
+                        {
+                            let mut writer = storage.write().expect("lock poisoning");
+                            writer.push(state.clone())?;
+                        }
+                        warn!(log, "New block saved"; "block" => format!("{:?}", block_hash));
+                        blocks.clear();
+                        block_storage.assign_to_context(block_hash, new_context_hash)?
+                    }
+                    ContextAction::Checkout { context_hash: _, .. } => {
+                        /**/
+                    }
+                    ContextAction::Mem { block_hash: Some(block_hash), .. }
                     | ContextAction::DirMem { block_hash: Some(block_hash), .. }
                     | ContextAction::Get { block_hash: Some(block_hash), .. }
                     | ContextAction::Fold { block_hash: Some(block_hash), .. } => {
                         let key = block_hash.clone();
                         let value = ContextRecordValue::new(msg);
                         context_storage.put(&key, &value)?;
                     }
-                    ContextAction::Commit { block_hash: Some(block_hash), new_context_hash, .. } => {
-                        block_storage.assign_to_context(block_hash, new_context_hash)?
-                    }
-                    ContextAction::Checkout { .. } => {
-                        // ...
-                    }
                     _ => (),
                 };
             }
             Err(err) => {
                 warn!(log, "Failed to receive event from protocol runner"; "reason" => format!("{:?}", err));
                 break;
-            },
+            }
         }
-
     }
 
     Ok(())
 }
 
+/// Get value contained in the dictionary, or create new entry associated with given key
+/// populated with default value
+fn get_default(state: &mut HashMap<Vec<u8>, Ctx>, block_hash: Vec<u8>) -> &mut Ctx {
```

Comment:
`get_default_WHAT` - use more descriptive function name

Reply:
`get(default)` and `set_default` are common method found on hashmap in other languages.
`get_default` -  Either gets value associated with the key or associate it with some default value and then returns it (in our case Default::default).

`set_default(k ,v)` - If key is present in hashmap, it works exactly as `get`. Otherwise it sets the key to the given value `v` first.